src/decoder.rs

```diff
@@ -86,36 +92,33 @@ impl<R: Read> Read for Decoder<R> {
     fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
         match self {
             Decoder::PlainText(decoder) => decoder.read(buf),
-            Decoder::Gzip(decoder) => decoder.read(buf).map_err(|e| {
-                if decoder.get_ref().has_errored {
-                    e
-                } else {
-                    io::Error::new(
-                        e.kind(),
-                        format!("error decoding gzip response body: {}", e),
-                    )
-                }
-            }),
-            Decoder::Deflate(decoder) => decoder.read(buf).map_err(|e| {
-                if decoder.get_ref().has_errored {
-                    e
-                } else {
-                    io::Error::new(
-                        e.kind(),
-                        format!("error decoding deflate response body: {}", e),
-                    )
-                }
-            }),
-            Decoder::Brotli(decoder) => decoder.read(buf).map_err(|e| {
-                if decoder.get_ref().has_errored {
-                    e
-                } else {
-                    io::Error::new(
-                        e.kind(),
-                        format!("error decoding brotli response body: {}", e),
-                    )
-                }
-            }),
+            Decoder::Gzip(decoder) => match decoder.read(buf) {
+                Ok(n) => Ok(n),
+                Err(e) if decoder.get_ref().has_errored => Err(e),
+                Err(_) if !decoder.get_ref().has_read_data => Ok(0),
```

Comment:
The check has to be done in this order to avoid ignoring `InnerReader` errors when no data has been read